src/libraries/Microsoft.PowerFx.Connectors/Execution/OpenApiJsonSerializer.cs

```diff
@@ -0,0 +1,201 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Dynamic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.Json;
+using Microsoft.OpenApi.Models;
+
+namespace Microsoft.PowerFx.Connectors.Execution
+{
+    internal class OpenApiJsonSerializer
+    {
+        private readonly Dictionary<string, object> _dic;
+        private readonly MemoryStream _stream;
+        private readonly Utf8JsonWriter _writer;
+        private readonly OpenApiSchema _schema;
+
+        internal OpenApiJsonSerializer(OpenApiSchema schema, Dictionary<string, object> dictionary)
+        {
+            _schema = schema;
+            _dic = dictionary;
+            _stream = new MemoryStream();
+            _writer = new Utf8JsonWriter(_stream, new JsonWriterOptions());
+        }
+
+        internal string ToJson()
+        {
+            _writer.WriteStartObject();
+            foreach (var prop in _schema.Properties)
+            {
+                WriteProperty(prop);
+            }
+
+            _writer.WriteEndObject();
+            _writer.Flush();
+            var json = Encoding.UTF8.GetString(_stream.ToArray());
+            return json;
+        }
+
+        private void WriteProperty(KeyValuePair<string, OpenApiSchema> property, object outerValue = null)
+        {
+            if (!_dic.TryGetValue(property.Key, out var value))
+            {
+                if (outerValue != null && outerValue is ExpandoObject eo)
+                {
+                    value = eo.FirstOrDefault(kvp => kvp.Key == property.Key);
+
+                    if (value != null && value is KeyValuePair<string, object> kvp)
+                    {
+                        value = kvp.Value;
+                    }
+                }
+            }
+
+            if (value == null)
+            {
+                throw new NotImplementedException($"Missing property {property.Key}, object is too complex or not supported");
+            }
+
+            switch (property.Value.Type)
+            {
+                case "array":
+                    _writer.WriteStartArray(JsonEncodedText.Encode(property.Key));
+                    if (value is not IEnumerable @enum)
+                    {
+                        throw new ArgumentException($"Type mismatch, expecting an array for {property.Key} and {value} is {value.GetType().FullName}");
+                    }
+
+                    foreach (var item in @enum)
+                    {
+                        WriteValue(item);
+                    }
+
+                    _writer.WriteEndArray();
+                    break;
+
+                case "null":
+                case "number":
+                case "boolean":
+                case "integer":
+                case "string":
+                    WriteProperty(property.Key, value);
```

Comment:
>WriteProperty [](http://example.com/codeflow?start=20&length=13)

These shouldn't all merge into the same case. 

Ie, you can do directly, like:
case "boolean":
   _writer.WriteBoolean(name, b);
   break;

In other words, there shouldn't be any possible path where "string" type end could potentially call WriterNumber().

Reply:
I think this would add complexity as the string here (null/number/boolean...) is the schema type and not the FormulaValue inner type. Some strings can "hide" a date for example. See [https://swagger.io/docs/specification/data-models/data-types/](https://swagger.io/docs/specification/data-models/data-types/)